File: CSED331/Assn4/2.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

class vertex_datum
{
public:
	int index;
	int rank;
	vertex_datum* parent;
	vertex_datum(int i)
	{
		rank = 0;
		index = i;
		parent = NULL;
	}
	void set_par(vertex_datum* p)
	{
		parent = p;
	}
};
class Edge
{
public:
	int s;
	int e;
	int w;
};

bool Sort_By_Weight(Edge l, Edge r) { return l.w < r.w; }

vertex_datum* Find(vertex_datum* x)
{
	if (x->parent == x)
		return x;
	else
	{
		x = Find(x->parent);
		return x;
	}
}

void Union(vertex_datum* x, vertex_datum* y)
{
	vertex_datum* root_x = Find(x);
	vertex_datum* root_y = Find(y);
	if (root_x->index == root_y->index)
		return;
	else
	{
		if (root_x->rank > root_y->rank)
		{
			root_y->parent = root_x;
		}
		else if(root_y->rank > root_x->rank)
		{
			root_x->parent = root_y;
		}
		else
		{
			root_x->parent = root_y;
			root_y->rank++;
		}
	}
}

int main()
{
	int t;
	cin >> t;
	for (int z = 0; z < t; z++)
	{
		int v, e;
		cin >> v >> e;
		vector<vertex_datum*> vertex(v);
		vector<Edge> edge(e);
		for (int i = 0; i < v; i++)
		{
			vertex[i] = new vertex_datum(i);
			vertex[i]->set_par(vertex[i]);
		}
		for (int i = 0; i < e; i++)
		{
			cin >> edge[i].s >> edge[i].e >> edge[i].w;
		}
		sort(edge.begin(), edge.end(), Sort_By_Weight);
		int weight = 0;
		for (int i = 0; i < e; i++)
		{
			int x, y;
			x = edge[i].s;
			y = edge[i].e;
			if (Find(vertex[x]) == Find(vertex[y]))
				continue;
			else
			{
				Union(vertex[x], vertex[y]);
				weight += edge[i].w;
			}
		}
		cout << weight << endl;



	}
	return 0;
}

File: CSED331/Assn4/2.py
class Union_Find:
    def init(self,index):
        self.index = index
        self.parent = index
        self.rank = 0
    

def find(x):
    if x.parent == x.index:
        return x.index
    else:
        x.parent = find(x.parent).index
        return x.parent

def Union(x,y):
    root_x = find(x)
    root_y = find(y)
    if root_x == root_y:
        return
    elif root_x.rank < root_y.rank:
        root_x.parent = root_y
        root_y.rank = root_x.rank + 1
    elif root_x.rank > root_y.rank:
        root_y.parent = root_x.index
        root_x.rank = root_y.rank + 1

t = int(input())

for z in range(t):
    v,e = map(int,input().split())
    vertex = []
    edge = []
    for i in range(v):
        t = Union_Find()
        t.init(i)
        vertex.append(t)
    for i in range(e):
        edge.append(list(map(int,input().split())))
    edge.sort(key = lambda x : (x[2]))
    weight = 0
    for i in range(e):
        print(find(vertex[edge[i][0]]),find(vertex[edge[i][1]]))
        if find(vertex[edge[i][0]]) == find(vertex[edge[i][1]]):
            continue
        else:
            weight += edge[i][2]
            Union(vertex[edge[i][0]],vertex[edge[i][1]])
    print(weight)
        
